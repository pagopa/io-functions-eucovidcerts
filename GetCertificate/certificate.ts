/* eslint-disable sort-keys */
/**
 * Certificate parsing utilities
 */
import { NonEmptyString } from "@pagopa/ts-commons/lib/strings";
import { Either, left } from "fp-ts/lib/Either";
import * as t from "io-ts";
import { WithinRangeInteger } from "@pagopa/ts-commons/lib/numbers";
import { DateFromString } from "@pagopa/ts-commons/lib/dates";
import { toTWithMap, toTWithMapOptional } from "../utils/conversions";
import { diseaseAgentTargeted } from "./valuesets/diseaseAgentTargeted";
import { vaccineProphylaxis } from "./valuesets/vaccineProphylaxis";
import { vaccineMedicinalProduct } from "./valuesets/vaccineMedicinalProduct";
import { marketingAuthorizationHolder } from "./valuesets/marketingAuthorizationHolders";
import { labTestTypes } from "./valuesets/labTestTypes";
import { labTestManufactorers } from "./valuesets/labTestManufactorers";
import { testResults } from "./valuesets/testResults";

type PersonName = t.TypeOf<typeof PersonName>;
const PersonName = t.interface({
  fn: t.string, // Familiy Name
  fnt: NonEmptyString, // Standardized Family Name
  gn: t.string, // Given Name
  gnt: t.string // Standardized Given Name
});

export type VaccinationEntry = t.TypeOf<typeof VaccinationEntry>;
export const VaccinationEntry = t.interface({
  tg: t.string.pipe(toTWithMap(diseaseAgentTargeted)), // disease or agent targeted
  vp: t.string.pipe(toTWithMap(vaccineProphylaxis)), // vaccine or prophylaxis
  mp: t.string.pipe(toTWithMap(vaccineMedicinalProduct)), // vaccine medicinal product
  ma: t.string.pipe(toTWithMap(marketingAuthorizationHolder)), // Marketing Authorization Holder
  dn: WithinRangeInteger(1, 9), // Dose Number
  sd: WithinRangeInteger(1, 9), // Total Series of Doses
  dt: DateFromString, // Date of Vaccination
  co: NonEmptyString, // Country of Vaccination
  is: NonEmptyString, // Certificate Issuer
  ci: NonEmptyString // Unique Certificate Identifier: UVCI
});

export type TestEntry = t.TypeOf<typeof TestEntry>;
export const TestEntry = t.intersection([
  t.interface({
    tg: t.string.pipe(toTWithMap(diseaseAgentTargeted)), // disease or agent targeted
    tt: t.string.pipe(toTWithMap(labTestTypes)), // covid-19 Lab Test Types
    sc: DateFromString, // Date/Time of Sample Collection
    tr: t.string.pipe(toTWithMap(testResults)), // Test Result
    tc: t.string, // Testing Centre
    co: t.string, // Country of Test
    is: t.string, // Issuer
    ci: t.string // Unique Certificate Identifier: UVCI
  }),
  t.partial({
    nm: t.string, // test name
    dr: DateFromString, // Date/Time of Test Result
    ma: t.string.pipe(toTWithMapOptional(labTestManufactorers)) // lab test manufactorers
  })
]);

export type RecoveryEntry = t.TypeOf<typeof RecoveryEntry>;
export const RecoveryEntry = t.intersection([
  t.interface({
    tg: t.string.pipe(toTWithMap(diseaseAgentTargeted)), // disease or agent targeted
    fr: DateFromString, // ISO 8601 Date of First Positive Test Result
    co: t.string, // Country of Test
    is: t.string, // Certificate Issuer
    df: DateFromString, // ISO 8601 Date: Certificate Valid From
    du: DateFromString, // ISO 8601 Date: Certificate Valid Until
    ci: t.string // Unique Certificate Identifier: UVCI
  }),
  t.partial({})
]);

export const VacCertificate = t.interface({
  ver: NonEmptyString,
  nam: PersonName,
  dob: DateFromString,
  v: t.readonlyArray<typeof VaccinationEntry>(VaccinationEntry)
});
export const TestCertificate = t.interface({
  ver: NonEmptyString,
  nam: PersonName,
  dob: DateFromString,
  t: t.readonlyArray<typeof TestEntry>(TestEntry)
});
export const RecoveryCertificate = t.interface({
  ver: NonEmptyString,
  nam: PersonName,
  dob: DateFromString,
  r: t.readonlyArray<typeof RecoveryEntry>(RecoveryEntry)
});

export type Certificates = t.TypeOf<typeof Certificates>;
export const Certificates = t.union([
  VacCertificate,
  TestCertificate,
  RecoveryCertificate
]);

/**
 * Parse a given base64 string representing a qrcode image to extract Certificate's informations
 *
 * @param qrcode
 * @returns either the Certificate object or a parsing error message
 */
export const parseQRCode = (_qrcode: string): Either<string, Certificates> =>
  left(`QRCode parsing not yet implemented`);
