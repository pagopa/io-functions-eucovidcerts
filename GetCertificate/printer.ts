/* eslint-disable sort-keys */
import {
  PreferredLanguage,
  PreferredLanguageEnum
} from "@pagopa/io-functions-commons/dist/generated/definitions/PreferredLanguage";
import * as o from "fp-ts/lib/Option";
import * as moment from "moment-timezone";
import { match } from "ts-pattern";
import { pipe } from "fp-ts/lib/function";
import { SupportedLanguage } from "../utils/conversions";
import {
  Certificates,
  TestCertificate,
  VacCertificate,
  VaccinationEntry,
  TestEntry,
  RecoveryCertificate,
  RecoveryEntry
} from "./certificate";
import * as vacDetailsEn from "./markdown/eucovidcertDetailsVaccinationEn";
import * as vacDetailsIt from "./markdown/eucovidcertDetailsVaccinationIt";
import * as vacDetailsDe from "./markdown/eucovidcertDetailsVaccinationDe";
import * as testDetailsIt from "./markdown/eucovidcertDetailsTestIt";
import * as testDetailsEn from "./markdown/eucovidcertDetailsTestEn";
import * as testDetailsDe from "./markdown/eucovidcertDetailsTestDe";
import * as recoveryDetailsIt from "./markdown/eucovidcertDetailsRecoveryIt";
import * as recoveryDetailsEn from "./markdown/eucovidcertDetailsRecoveryEn";
import * as recoveryDetailsDe from "./markdown/eucovidcertDetailsRecoveryDe";
import * as vacInfoMultilanguage from "./markdown/eucovidcertInfoVaccinationEn";
import * as vacInfoGerman from "./markdown/eucovidcertInfoVaccinationDe";
import * as expiredInfoIt from "./markdown/eucovidcertExpiredInfoIt";
import * as expiredInfoEn from "./markdown/eucovidcertExpiredInfoEn";
import * as expiredInfoDe from "./markdown/eucovidcertExpiredInfoDe";

import { labTestTypes, molecularTest } from "./valuesets/labTestTypes";

const TIME_ZONE = "Europe/Rome";

const DATE_FORMAT_EN = "YYYY-MM-DD";
const DATE_FORMAT_ITA = "DD-MM-YYYY";

const dateFormatForLanguage: { [key in SupportedLanguage]: string } = {
  [PreferredLanguageEnum.it_IT]: DATE_FORMAT_ITA,
  [PreferredLanguageEnum.en_GB]: DATE_FORMAT_EN,
  [PreferredLanguageEnum.de_DE]: DATE_FORMAT_EN
};

const DATE_TIME_FORMAT_EN = "YYYY-MM-DD HH:mm";
const DATE_TIME_FORMAT_ITA = "DD-MM-YYYY HH:mm";

const dateAndTimeFormatForLanguage: { [key in SupportedLanguage]: string } = {
  [PreferredLanguageEnum.it_IT]: DATE_TIME_FORMAT_ITA,
  [PreferredLanguageEnum.en_GB]: DATE_TIME_FORMAT_EN,
  [PreferredLanguageEnum.de_DE]: DATE_TIME_FORMAT_EN
};
interface IPrintersForLanguage {
  readonly expiredInfoPrinter: () => string;
  readonly detailVaccinePrinter: (v: VaccinationEntry) => string;
  readonly detailTestPrinter: (t: TestEntry) => string;
  readonly detailRecoveryPrinter: (r: RecoveryEntry) => string;
  readonly infoVaccinePrinter: (c: Certificates) => string;
  readonly infoTestPrinter: (c: Certificates) => string;
  readonly infoRecoveryPrinter: (c: Certificates) => string;
}

const printersConfigurations: {
  [key in SupportedLanguage]: IPrintersForLanguage;
} = {
  [PreferredLanguageEnum.it_IT]: {
    expiredInfoPrinter: expiredInfoIt.getInfoPrinter,
    detailVaccinePrinter: vacDetailsIt.getDetailPrinter,
    detailTestPrinter: testDetailsIt.getDetailPrinter,
    detailRecoveryPrinter: recoveryDetailsIt.getDetailPrinter,
    infoVaccinePrinter: vacInfoMultilanguage.getInfoPrinter,
    infoTestPrinter: vacInfoMultilanguage.getInfoPrinter,
    infoRecoveryPrinter: vacInfoMultilanguage.getInfoPrinter
  },
  [PreferredLanguageEnum.en_GB]: {
    expiredInfoPrinter: expiredInfoEn.getInfoPrinter,
    detailVaccinePrinter: vacDetailsEn.getDetailPrinter,
    detailTestPrinter: testDetailsEn.getDetailPrinter,
    detailRecoveryPrinter: recoveryDetailsEn.getDetailPrinter,
    infoVaccinePrinter: vacInfoMultilanguage.getInfoPrinter,
    infoTestPrinter: vacInfoMultilanguage.getInfoPrinter,
    infoRecoveryPrinter: vacInfoMultilanguage.getInfoPrinter
  },
  [PreferredLanguageEnum.de_DE]: {
    expiredInfoPrinter: expiredInfoDe.getInfoPrinter,
    detailVaccinePrinter: vacDetailsDe.getDetailPrinter,
    detailTestPrinter: testDetailsDe.getDetailPrinter,
    detailRecoveryPrinter: recoveryDetailsDe.getDetailPrinter,
    infoVaccinePrinter: vacInfoGerman.getInfoPrinter,
    infoTestPrinter: vacInfoGerman.getInfoPrinter,
    infoRecoveryPrinter: vacInfoGerman.getInfoPrinter
  }
};

export const defaultPrinter: IPrintersForLanguage =
  printersConfigurations[PreferredLanguageEnum.en_GB];

/**
 * Return the correct IPrintersForLanguage based on preferred language,
 * or default IPrintersForLanguage if no language is selected
 *
 * @param lang the preferred language, if any
 * @returns
 */
export const getPrinterForLanguage = (
  lang: o.Option<PreferredLanguage>
): IPrintersForLanguage =>
  pipe(
    lang,
    o.chain(l =>
      o.fromNullable(
        (printersConfigurations as Record<
          PreferredLanguage,
          IPrintersForLanguage
        >)[l]
      )
    ),
    o.getOrElse(() => defaultPrinter)
  );

/**
 * Returns the Detail markdown filled with data from the input Certificate
 * based on preferred language and the Certificate type (vac, test, rec)
 *
 * @param lang the preferred language
 * @param c the Certificate
 * @returns a string containing the Info markdown
 */
export const printDetails = (
  lang: o.Option<PreferredLanguage>,
  c: Certificates
): string =>
  match(c)
    .when(VacCertificate.is, vc =>
      getPrinterForLanguage(lang).detailVaccinePrinter(vc.v[0])
    )
    .when(TestCertificate.is, tc =>
      getPrinterForLanguage(lang).detailTestPrinter(tc.t[0])
    )
    .when(RecoveryCertificate.is, rc =>
      getPrinterForLanguage(lang).detailRecoveryPrinter(rc.r[0])
    )
    .exhaustive();

/**
 * Returns the Info markdown filled with data from the input Certificate
 * based on preferred language
 *
 * @param lang the preferred language
 * @param c the Certificate
 * @returns a string containing the Info markdown
 */
export const printInfo = (
  lang: o.Option<PreferredLanguage>,
  c: Certificates
): string => getPrinterForLanguage(lang).infoVaccinePrinter(c);

/**
 * Returns UVCI of the first certificate (either Vaccination, Test or Recovery)
 *
 * @param _lang the preferred language (unused)
 * @param c the Certificate
 * @returns the UVCI
 */
export const printUvci = (
  _lang: o.Option<PreferredLanguage>,
  c: Certificates
): string =>
  match(c)
    .when(VacCertificate.is, cv => cv.v[0].ci)
    .when(TestCertificate.is, ct => ct.t[0].ci)
    .when(RecoveryCertificate.is, cr => cr.r[0].ci)
    .exhaustive();

/**
 * Returns the markdown info text to be rendered when a Certificate is found to be expired
 * based on preferred language
 *
 * @param lang the preferred language
 * @returns a string containing the Info markdown
 */
export const printExpiredInfo = (lang: o.Option<PreferredLanguage>): string =>
  getPrinterForLanguage(lang).expiredInfoPrinter();

/**
 * Check if Vaccination has ended
 *
 * @param v the VaccinationEntry to check
 * @returns true if process ended, false otherwise
 */
export const isVaccinationProcessEnded = (v: VaccinationEntry): boolean =>
  v.dn >= v.sd; // we may have more doses than the total in case of additional, "booster doses"

/**
 * Test validity in hours, based on type
 *
 * @param v the VaccinationEntry to check
 * @returns the validity of the test, in hours
 */
export const testValidity = (v: TestEntry): string =>
  v.tt === labTestTypes[molecularTest] ? "72" : "48";

/**
 * Format date value based on language
 *
 * @param d the date
 * @param _lang the preferred language
 * @returns a formatted date
 */
export const formatDate = (d: Date, lang: SupportedLanguage): string =>
  pipe(dateFormatForLanguage[lang], format =>
    moment(d)
      .tz(TIME_ZONE)
      .format(format)
  );

/**
 * Format date and time value based on language
 *
 * @param d the date
 * @param _lang the preferred language
 * @returns a formatted date with time
 */
export const formatDateAndTime = (d: Date, lang: SupportedLanguage): string =>
  // eslint-disable-next-line sonarjs/no-identical-functions
  pipe(dateAndTimeFormatForLanguage[lang], format =>
    moment(d)
      .tz(TIME_ZONE)
      .format(format)
  );

/**
 * The issuer (is) field value used by Italy Healthcare Department
 */
export const ITALY_HEALTHCARE_ISSUER = "IT";

export const HEALTHCARE_DEP_IT = "Ministero della Salute";
export const HEALTHCARE_DEP_EN = "Ministry of Health";
export const HEALTHCARE_DEP_DE = "Gesundheitsministerium";

const healthcareIssuerForLanguage: { [key in SupportedLanguage]: string } = {
  [PreferredLanguageEnum.it_IT]: HEALTHCARE_DEP_IT,
  [PreferredLanguageEnum.en_GB]: HEALTHCARE_DEP_EN,
  [PreferredLanguageEnum.de_DE]: HEALTHCARE_DEP_DE
};

/**
 * Format Certificate Issuer based on its value and preferred language
 *
 * @param c the issuer certificate value
 * @param _lang the preferred language
 * @returns a formatted Certificate Issuer string
 */
export const formatCertificateIssuer = (
  c: string,
  lang: SupportedLanguage
): string =>
  c !== ITALY_HEALTHCARE_ISSUER ? c : healthcareIssuerForLanguage[lang];
