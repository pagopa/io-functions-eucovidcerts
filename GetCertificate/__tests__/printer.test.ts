import {
  PreferredLanguage,
  PreferredLanguageEnum
} from "@pagopa/io-functions-commons/dist/generated/definitions/PreferredLanguage";
import { some } from "fp-ts/lib/Option";

import { Certificates } from "../certificate";
import { printDetails, printInfo, formatCertificateIssuer } from "../printer";

describe("Printer", () => {
  it("should format Certificate Issuer with Ministero della salute if value is IT and language it_IT", () => {
    const result = formatCertificateIssuer("IT", PreferredLanguageEnum.it_IT);
    expect(result).toEqual("Ministero della Salute");
  });
  it("should format Certificate Issuer with Ministry of Health if value is IT and language it_IT", () => {
    const result = formatCertificateIssuer("IT", PreferredLanguageEnum.en_GB);
    expect(result).toEqual("Ministry of Health");
  });
  it("should format Certificate Issuer with its value if value is different from IT and language it_IT", () => {
    const result = formatCertificateIssuer(
      "a value",
      PreferredLanguageEnum.en_GB
    );
    expect(result).toEqual("a value");
  });
});
