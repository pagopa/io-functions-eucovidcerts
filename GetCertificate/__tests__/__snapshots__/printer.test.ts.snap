// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Printers should print it markdown - test certificate 1`] = `
"## Certificato di Test
***
**Certificazione valida 48 ore dall'ora del prelievo**
***

Malattia o agente bersaglio  
**COVID-19**

Tipo di test  
**Nucleic acid amplification with probe detection**

Nome del test molecolare
**Roche LightCycler qPCR**




Data e ora del prelievo del campione  
**2021-4-1**

Data e ora del risultato del test
**2021-4-2**

Risultato del test  
**Negativo**

Centro o struttura in cui è stato eseguito il test  
**Policlinico Umberto I**

Stato Membro in cui è stato eseguito il test  
**IT**

Struttura che ha rilasciato la certificazione  
**IT**

Identificativo univoco del certificato  
**01IT0BFC9866D3854EAC82C21654B6F6DE32#1**
  
***

*Questo certificato non è un documento di viaggio. 
Le evidenze scientifiche sulla vaccinazione, sui test e sulla guarigione da COVID-19 continuano ad evolversi, anche in considerazione delle nuove varianti del virus.*  
*Prima di viaggiare, si prega di controllare le misure di salute pubblica applicate nel luogo di destinazione e le relative restrizioni anche consultando il sito:*   
[https://reopen.europa.eu/it](https://reopen.europa.eu/it)
  
***
  
Per ulteriori informazioni e per scaricare il certificato in formato PDF stampabile vai su 
[www.dgc.gov.it](https://www.dgc.gov.it)"
`;

exports[`Printers should print it markdown - vaccine certificate 1`] = `
"## Certificato di Vaccinazione
***
**Certificazione valida 270 giorni (9 mesi) dalla data dell'ultima somministrazione**
***

Malattia o agente bersaglio  
**COVID-19**

Tipo di vaccino somministrato  
**SARS-CoV-2 mRNA vaccine**

Denominazione del vaccino  
**Comirnaty**

Produttore o titolare dell’AIC del vaccino  
**Biontech Manufacturing GmbH**

Numero della dose effettuata / numero totale dosi previste  
**2 / 2**

Data dell’ultima somministrazione (yyyy-mm-dd) 
**2021-3-6**

Stato Membro in cui è stata eseguita la vaccinazione    
**IT**

Struttura che ha rilasciato la certificazione 
**IT-issuer**

Identificativo univoco del certificato  
**01ITE7300E1AB2A84C719004F103DCB1F70A#6**
  
***

*Questo certificato non è un documento di viaggio. 
Le evidenze scientifiche sulla vaccinazione, sui test e sulla guarigione da COVID-19 continuano ad evolversi, anche in considerazione delle nuove varianti del virus.*  
*Prima di viaggiare, si prega di controllare le misure di salute pubblica applicate nel luogo di destinazione e le relative restrizioni anche consultando il sito:*   
[https://reopen.europa.eu/it](https://reopen.europa.eu/it)
  
***
  
Per ulteriori informazioni e per scaricare il certificato in formato PDF stampabile vai su 
[www.dgc.gov.it](https://www.dgc.gov.it)"
`;
